### PR TITLE
Fix behavior of instmap.add when partial function is specified

### LINKED ISSUE #6278: InstructionScheduleMap.add incorrectly determines function signature
### Information

- **Qiskit Terra version**: '0.18.0.dev0+98130dd'
- **Python version**: 3.8
- **Operating system**: Windows

### What is the current behavior?

The `InstructionScheduleMap` allows to add callable schedules. This works for normal callables, but fails for callables defined with `partial`.

### Steps to reproduce the problem

A minimal example is:
```
from functools import partial
from qiskit import *
from qiskit.test.mock import FakeAthens
from qiskit.circuit import Gate, Parameter
from qiskit.pulse import Constant, DriveChannel
                          
backend = FakeAthens()
defaults = backend.defaults()
inst_map = defaults.instruction_schedule_map

def build_stage_schedule(duration, name, amp, frequency_offset, channel):
    with pulse.build() as pulse_prog:
        with pulse.frequency_offset(frequency_offset, channel):
            pulse.play(pulse.library.Constant(duration=duration, amp=amp), channel)
    return pulse_prog

frequency=Parameter('f')

def callable_schedule(f):
    schedule = build_stage_schedule(4, 'A', .1, f, DriveChannel(0))
    return schedule

inst_map.add('mygate', [0], callable_schedule, [frequency.name]) # works

def callable_schedule2(f, amplitude):
    schedule = build_stage_schedule(4, 'A', amplitude, f, DriveChannel(0))
    return schedule

inst_map.add('mygate', [0], partial(callable_schedule2, amplitude=.4), [frequency.name]) # fails
````

### What is the expected behavior?

Adding the second callable schedule `partial(callable_schedule2, amplitude=.4)` should have worked.

### Suggested solutions

Modify the line `func_parameters = list(inspect.signature(schedule).parameters.keys())` from the `add` method so that the correct signature is used.

### PR DESCRIPTION
**Pull request by nkanazawa1989**
### Summary
Fix #6278 

### Details and comments
See unittest for details

### CODE REVIEW

**Comment by taalexander** [comment]
I don't think special-casing `partial` support is the right way to handle this. I wrote up a long response and posted it to #6278 the other day but it seems to not have been posted ☹️. The gist of the problem is that it seems [partial](https://docs.python.org/3/library/functools.html#functools.partial) rather than binding supplied arguments fully, it creates them as keyword arguments in the returned partial object.

```
>>> functools.partialmethod(<function callable_schedule2 at 0x7ff790b79040>, , amplitude=0.4)
```

```python
inspect.signature(partial(callable_schedule2, amplitude=.4))
```

```
>>> <Signature (f, *, amplitude=0.4)>
```

The signature above is clearly not supported by `InstructionScheduleMap` as it currently only supports functions with a fixed number of arguments. The solution in this PR special cases `partial` but this is a more general policy of how keyword arguments should be handled in the function. 

There are several options:
1. Treat them as required arguments for scheduling (current behavior)
2. Ignore them as arguments (easy fix may be undesirable)
3. Special case `partial` (this PR, possibly undesirable as we will run into this problem again with similar function signature situations. We should define our policy).
4. Treat them as args based on ordering (would be slightly more effort, but could be done by storing signature objects rather than just a tuple of args?).

Right now I lean towards require signature being args only as Qiskit circuits/gates do not support variadic arguments. However, I am open to others' inputs.

**Comment by peendebak** [comment]
@taalexander You accidentality posted your comment to https://github.com/Qiskit/qiskit-terra/issues/5951. 

I can see that 2., 3. or 4. might have some complications in actual implementation. If you decide to stick with 1., then perhaps document this and offer some method to update the signature of a `partial`. E.g., 
```
inspect.signature(partial(callable_schedule2, amplitude=.4))
>>> <Signature (f, *, amplitude=0.4)>
inspect.signature(only_args_signature(partial(callable_schedule2, amplitude=.4)))
>>> <Signature (f )>
```
I could not easily find an option for such a `only_args_signature`

**Comment by taalexander** [comment]
> @taalexander You accidentality posted your comment to https://github.com/Qiskit/qiskit-terra/issues/5951. 
> 
> I can see that 2., 3. or 4. might have some complications in actual implementation. If you decide to stick with 1., then perhaps document this and offer some method to update the signature of a `partial`. E.g., 
> ```
> inspect.signature(partial(callable_schedule2, amplitude=.4))
> >>> <Signature (f, *, amplitude=0.4)>
> inspect.signature(only_args_signature(partial(callable_schedule2, amplitude=.4)))
> >>> <Signature (f )>
> ```
> I could not easily find an option for such a `only_args_signature`

Thanks for tracking that down 😌. I agree, either a solution or documentation is necessary. I also couldn't find an easy way around this with `partial`.

**Comment by nkanazawa1989** [comment]
I see @taalexander 's point. I carefully investigated [inspect](https://docs.python.org/3/library/inspect.html) module and found that this module covers all of those complicated signatures. So I changed `.add` logic to internally generate `Signature` object for schedule sub-classes, and offloaded parameter management to this object.

File: qiskit/pulse/instruction_schedule_map.py [hunk 1]
```diff
@@ -235,44 +236,51 @@ def add(self,
...
+                    annotation=ParameterValueType,
+                    kind=inspect.Parameter.POSITIONAL_OR_KEYWORD
+                )
+                parameters.append(param_signature)
+            signature = inspect.Signature(parameters=parameters, return_annotation=Schedule)
+
+        elif callable(schedule):
+            if arguments:
+                warnings.warn('Arguments are overridden by the callback function signature. '
+                              'Input `arguments` is just ignored.', UserWarning)
```
Comment (taalexander) [inline qiskit/pulse/instruction_schedule_map.py:276]:
```suggestion
                              'Input `arguments` are ignored.', UserWarning)
```

File: qiskit/pulse/instruction_schedule_map.py [hunk 2]
```diff
@@ -235,44 +236,51 @@ def add(self,
...
-        if not (isinstance(schedule, (Schedule, ScheduleBlock)) or callable(schedule)):
-            raise PulseError('Supplied schedule must be either a Schedule, ScheduleBlock or a '
-                             'callable that outputs a schedule.')
-
-        # initialize parameter list
-        if callable(schedule):
-            func_parameters = list(inspect.signature(schedule).parameters.keys())
         else:
-            func_parameters = set(param.name for param in schedule.parameters)
+            raise PulseError('Supplied schedule must be one of the Schedule, ScheduleBlock and '
```
Comment (taalexander) [inline qiskit/pulse/instruction_schedule_map.py:280]:
```suggestion
            raise PulseError('Supplied schedule must be one of the Schedule, ScheduleBlock or a '
```

File: releasenotes/notes/fix-instmap-add-partial-bind-604b74158e5a56e3.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+fixes:
+  - |
+    Fix argument parsing bug of :meth:`InstructionScheduleMap.add` when
+    `functools.partial` function is input.
```
Comment (taalexander) [inline releasenotes/notes/fix-instmap-add-partial-bind-604b74158e5a56e3.yaml:5]:
```suggestion
    Fix argument parsing bug of :meth:`InstructionScheduleMap.add` when a
    `functools.partial` function is input.
```

**Review by taalexander — CHANGES_REQUESTED** [review]
This is excellent @nkanazawa1989, I really appreciate you diving into this and coming up with an overall better implementation and great solution 😄 . Just a few small changes requested.

**Review by taalexander — APPROVED** [review]

**Pull request merged**